sensor: Add NTC thermistor sensor for TDK Robokit1 board

asemjonovs opened the pull request:
The TDK Robokit1 board has an `epcos,b57861s0103a039` NTC thermistor.  
This exists on the SPI bus and is read via the `ti,ads7052` ADC

asemjonovs committed:
adc: Add TI ADS7052 SPI driver

Add driver for TI ADS7052.

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

asemjonovs committed:
sensor: Generic driver for NTC Thermistor

Driver for NTC Thermistors attached to ADC

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

asemjonovs committed:
tdk_robokit1: Add alias for thermometer sample app

Add ambient-temp0 alias to verify via sample app

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

asemjonovs commented:
The NTC thermistor has a lookup table to translate the resistance to a temperature value.  Do we want this table to exist in c code or in the device tree?

yperess commented on drivers/sensor/b57861s0103a039/b57861s0103a039.c:
```diff
@@ -0,0 +1,330 @@
+	struct adc_sequence sequence = {.options = NULL,
+					.buffer = &data->raw,
+					.buffer_size = sizeof(data->raw),
+					.calibrate = false};
```
nit: add trailing `,` and run clang-format

yperess commented on drivers/sensor/b57861s0103a039/b57861s0103a039.c:
```diff
@@ -0,0 +1,330 @@
+		temp = get_temp_mc(dev->data, ohm);
+		val->val1 = temp / 1000;
+		val->val2 = (temp % 1000) * 1000;
+		LOG_ERR("ntc temp says %u", val->val1);
```
LOG_ERR?

yperess commented on drivers/sensor/b57861s0103a039/b57861s0103a039.c:
```diff
@@ -0,0 +1,330 @@
+	return 0;
+}
+
+static int b57861s0103a039_attr_set(const struct device *dev, enum sensor_channel chan,
+				    enum sensor_attribute attr, const struct sensor_value *val)
+{
+	return -ENOTSUP;
+}
+
+static int b57861s0103a039_attr_get(const struct device *dev, enum sensor_channel chan,
+				    enum sensor_attribute attr, struct sensor_value *val)
+{
+	return -ENOTSUP;
+}
```
If these aren't needed you can just leave them out of the API, the sensor.h static inline functions will check these and return -ENOTSUP for you

yperess commented on drivers/sensor/b57861s0103a039/b57861s0103a039.c:
```diff
@@ -0,0 +1,330 @@
+	data->comp_table = &b57861s0103a039_type;
+
+	if (!device_is_ready(cfg->adc_channel.dev)) {
+		printk("ADC controller device is not ready\n");
```
nit: `LOG_ERR` (same below)

yperess commented on drivers/sensor/b57861s0103a039/b57861s0103a039.c:
```diff
@@ -0,0 +1,330 @@
+
+int b57861s0103a039_init(const struct device *dev)
+{
+	if (b57861s0103a039_reset(dev)) {
+		LOG_ERR("could not initialize sensor");
+		return -EIO;
+	}
+
+	return 0;
```
Can we just call `return b57861s0103a039_reset(dev);`?

yperess reviewed (changes requested):
Can you also add `boards/tdk_robokit1.overlay` to the sensor_shell sample?

teburd commented on drivers/sensor/b57861s0103a039/b57861s0103a039.c:
```diff
@@ -0,0 +1,313 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT epcos_b57861s0103a039_spi
```
The file name and very long compat are pretty cryptic, can we abbreviate this a bit to epcos_b57861s.c with perhaps some DT attributes to help clarify which variant of the part family it is. E.g. the 100k/10k/etc part. Some parts will likely be exactly the same but qualified for auto and such.

teburd commented on dts/bindings/sensor/epcos,b57861s0103a039.yaml:
```diff
@@ -0,0 +1,37 @@
+
+description: EPCOS, B57861S0103A039 NTC Thermistor
+
+compatible: "epcos,b57861s0103a039-spi"
```
Why is spi needed in the compatible name? The thermistor could be connected to any adc presumably, including the builtin soc adc?

teburd commented on tests/drivers/build_all/sensor/spi.dtsi:
```diff
@@ -337,3 +337,32 @@ test_spi_icm426888: icm42688@28 {
+
+#include <zephyr/dt-bindings/adc/adc.h>
+
+test_spi_ads7052: ads7052@29 {
```
This should go under the adc build all instead

teburd commented on tests/drivers/build_all/sensor/spi.dtsi:
```diff
@@ -337,3 +337,32 @@ test_spi_icm426888: icm42688@28 {
 	gyro-fs = <2000>;
 };
+
+#include <zephyr/dt-bindings/adc/adc.h>
```
Moving the below adc to the correct place, this shouldn't be needed

teburd commented on tests/drivers/build_all/sensor/spi.dtsi:
```diff
@@ -337,3 +337,32 @@ test_spi_icm426888: icm42688@28 {
+	};
+};
+
+test_spi_b57861s0103a039: b57861s0103a039@2a {
```
This should probably go into a new adc.dtsi which is included in app.overlay, @MaureenHelm thoughts?

teburd reviewed (changes requested):
This is quite a PR for adding the ntc thermocouple, I can't comment too much on the ADC stuff as I'm not as familiar with the API and implementations. To me it looks ok. A little bit of a bummer yet another thread is needed there to go fetch a sample over SPI.

The things I'd like to see changed for a +1 from me is really around organization and naming.

Perhaps placing the driver in an epcos_ntc or tdk_ntc directory would help identify it. This along with the driver compat, file name and function prefixes shortened to the part family (b57861s) might be sensible and make it easier to recognize. At the moment the driver name, compat and prefixes are very long giving few clues.

DT attributes could be used to setup the needed variant options at compile time perhaps then perhaps to account for differences in lookup tables for interpolating over. Anything to help avoid duplicating work if someone goes to add another variant of the part to the tree would be useful.

As noted in some comments this isn't a spi connected sensor, its an adc connected sensor and should be setup as such.

teburd commented on tests/drivers/build_all/sensor/spi.dtsi:
```diff
@@ -343,3 +343,14 @@ test_spi_max31855: max31855@29 {
 	spi-max-frequency = <0>;
 };
+
+test_spi_b57861s: b57861s@2a {
```
This should probably be test_adc_b57861 and in the build_all/sensor/adc.dtsi

teburd reviewed (changes requested):
the dtsi for the build all still doesn't look quite right to me, the rest looks much nicer

MaureenHelm commented on drivers/sensor/b57861s/Kconfig:
```diff
@@ -0,0 +1,14 @@
+	bool "EPCOS B57861S NTC Thermistor sensor"
+	default y
+	depends on DT_HAS_EPCOS_B57861S_ENABLED
+	select SPI
```
Remove this and let the ADC driver select SPI.

MaureenHelm commented on drivers/sensor/b57861s/b57861s.c:
```diff
@@ -0,0 +1,316 @@
+ */
+
+/* clang-format off */
+static const struct ntc_compensation b57861s0103a039[] = {
```
Fix this long name too

MaureenHelm commented on samples/sensor/thermometer/boards/tdk_robokit1.overlay:
```diff
@@ -0,0 +1,13 @@
+ * Copyright (c) 2022, Gerson Fernando Budke
+ */
+
+#include <zephyr/dt-bindings/spi/spi.h>
```
Why?

asemjonovs commented on samples/sensor/thermometer/src/main.c:
```diff
@@ -12,15 +12,21 @@ void main(void)
-	printf("Thermometer Example! %s\n", CONFIG_ARCH);
+	printk("Thermometer Example! %s\n", CONFIG_ARCH);
 
+#if DT_NODE_EXISTS(DT_PATH(zephyr_user)) && \
+	DT_NODE_HAS_PROP(DT_PATH(zephyr_user), sensor)
+	temp_dev = DEVICE_DT_GET(DT_PROP(DT_PATH(zephyr_user), sensor));
+#else
 	temp_dev = device_get_binding("TEMP_0");
+#endif
```
Hi Maureen - I updated this to match how `samples/drivers/adc` works using the `zephyr,user` device tree property for specific boards to specify which device to use.  I left `device_get_binding` to continue to support existing behavior.  I can remove this so future development is forced to use `zephyr,user` node

yperess commented on drivers/adc/adc_ads7052.c:
```diff
@@ -0,0 +1,309 @@
+	k_sem_init(&data->sem, 0, 1);
+
+	if (!spi_is_ready_dt(&config->bus)) {
+		printk("SPI bus %s not ready", config->bus.bus->name);
```
LOG_ERR

yperess reviewed (changes requested):
super minor nit, but LGTM otherwise

cfriedt commented on drivers/sensor/b57861s/b57861s.c:
```diff
@@ -0,0 +1,316 @@
+	enum ntc_type_e connection_type;
+};
+
+static void lookup_comp(const struct ntc_type *data, unsigned int ohm, int *i_low, int *i_high)
```
Could `bsearch()` help here?

asemjonovs replied:
`bsearch` looks to find an exact match and returns a pointer to it, if not found returns NULL.  
While `lookup_comp` performs a binary search, its most unlikely to find an ohm value that matches.  It finds the `low` and `high` indexes where `ohm` falls between.  I might be able to make the compare function passed to `bsearch` to return found in these scenarios but seems a little hacky to me as will need to check the pointer returned then check if it's upper or lower bound.

cfriedt replied:
https://pastebin.com/prpRVmjn

cfriedt replied:
That's cool. If you would consider another generic tweak, you could put the table into Device Tree 😃

cfriedt reviewed (changes requested):
Looks like the standard libc bsearch could be used (space optimization).

teburd reviewed (changes requested):
Minor copyright nit

cfriedt commented on drivers/sensor/b57861s/b57861s.c:
```diff
@@ -0,0 +1,316 @@
+		 * data->comp[start].ohm >= ohm >= data->comp[end].ohm
+		 */
+	}
+	/*uint64_t
```
Is the `uint64_t` a typo?

MaureenHelm commented on dts/bindings/sensor/ntc_thermistor.yaml:
```diff
@@ -0,0 +1,37 @@
+
+description: Common properties for thermistors
+
+compatible: "zephyr,thermistor"
```
The filename needs to match the compatible

MaureenHelm commented on dts/bindings/sensor/ntc_thermistor.yaml:
```diff
@@ -0,0 +1,37 @@
+    temp-c:
+      type: int
+      required: true
+      description: Temperature (Celcius) in a sample
```
Celsius

MaureenHelm reviewed (changes requested):
Couple of nits, but otherwise looks great!

mbolivar-nordic commented on dts/bindings/sensor/zephyr,thermistor.yaml:
```diff
@@ -0,0 +1,37 @@
+    description: Number of sample data points for linear interpolation
+
+child-binding:
+  description: >
```
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#descriptions

mbolivar-nordic commented on dts/common/zephyr_thermistor.dtsi:
```diff
@@ -0,0 +1,98 @@
+ */
+
+/ {
+	/omit-if-no-ref/ thermistor_R25_10K_B_3974: thermistor-R25-10K-B-3974 {
```
What is going on here? Why is this being added to `dts/common/`? This all looks like very hardware-specific property settings that don't belong here.

asemjonovs replied:
Hi Marti - I'm not exactly sure where this should live.  This looked like a good place.  NTC thermistors can share similar characteristics so the table can be used in other places (why I chose `dts/common`), so another manufacturer can reference this same table.  Would creating a `dts/thermistor` folder be more appropriate?

mbolivar-nordic replied:
> Would creating a dts/thermistor folder be more appropriate?

I don't think so, no. Look at the other directories under `dts`. We currently have SoC specific DTSI files and bindings in here, besides the common directory which contains things that apply to virtually every .dts used by zephyr. Putting a random node in here that looks like a hardware-on-my-desk type thing (the `_R25_10K_B_3974` is a tell, for example, at least if I'm reading that correctly) doesn't make sense IMO.

> can be used in other places

where?

cfriedt replied:
IIRC, it's possible to have many device tree compatibles all reference the same binding, and if the table is const in DT, then it might be possible to even have 1 driver for serval different thermistor profiles too.

Marti us the expert there though, so I defer to his advice.

mbolivar-nordic replied:
> A search on digikey came up with 52 [thermistors](https://www.digikey.com/en/products/filter/temperature-sensors/ntc-thermistors/508?s=N4IgjCBcoBwGwGYqgMZQGYEMA2BnApgDQgD2UA2iAEwAMYVA7FSALrEAOALlCAMqcAnAJYA7AOYgAvsXhxkINJCx4ipCiAYIALDSRsQXHv2HipkyUA) that have the same electrical characteristics that can reference this same table.

Just to be clear, I fully recognize that you are the domain expert on thermistors here, and I believe you that this is a frequently seen piece of hardware.

> So being consistent with other *.dtsi files, I think this file should go under dts somewhere

However, I do not yet agree that this approach is consistent with other *.dtsi files, or that it should go under dts/. Let me try to explain and we'll see how we can get this done.

Take ns16550 UARTs for example. This is also a very common piece of hardware. As far as I know, we do not have a *.dtsi file with a "reference" ns16550 node inside it anywhere under dts/, and I would argue that we should not. Would you want that to be added to a bunch of devicetrees that don't need it, and then ripped out with `/omit-if-no-ref/` in 99.999...% of cases? I think that would be inappropriate. Reasoning by analogy, I think this current approach is also inappropriate.

> may be a new use case for a device tree node that's shared between thermistors, that's not part of an SOC/board.

The usual idiom is to provide example nodes in the `description:` field of the devicetree binding when there are common examples. That lets everyone who has that piece of hardware see a working example they can copy/paste into their SoC DTSI or board DTS. It really doesn't seem like this is any different.

>> can be used in other places
>
> where?

That is why I am asking you where specifically these "other" places are. If it's just board DTS files, for example, IMO we should go with the usual idiom of examples in the description. If this really is some sort of rare bird, however, I think it's fair to ask you to provide more examples of where it's going to show up that prove the point.

asemjonovs replied:
Hi Marti - I'm discussing with my team on this.  You can disregard my most recent push moving this file out of `dts/common`.

asemjonovs replied:
Hi @mbolivar-nordic, @cfriedt 

What do you think of defining the table in C code guarded by `DT_HAS_COMPAT_STATUS_OKAY` check?

Something like

zephyr_thermistor.c
```
#if DT_HAS_COMPAT_STATUS_OKAY(zephyr_thermistor_rt_table)
struct ntc_compensation THERMISTOR_NAME = [...];
#endif 
```

boards/arm/tdk_robokit1/tdk_robokit1-common.dtsi
```
	thermistor_R25_10K_B_3974: thermistor_R25_10K_B_3974 {
		compatible = "zephyr,thermistor-rt-table";
	};
	temp_sensor: ambient_temp_sensor {
		compatible = "zephyr,thermistor";
		io-channels = <&spi_adc 0>;
		thermistor = <&thermistor_R25_10K_B_3974>;
		r25-ohm = <10000>;
		pullup-mv = <3300>;
		pullup-ohm = <0>;
		pulldown-ohm = <10000>;
		connection-type = "NTC_CONNECTED_POSITIVE";
	};
```

mbolivar-nordic replied:
@asemjonovs I really can't say without knowing more about the places this is going to be used like I was asking about. The heart of my questions/concerns here are "why is this hardware so special that it's different than literally every other situation we have in the tree", which your most recent question doesn't answer for me. Make sense?

asemjonovs replied:
@mbolivar-nordic - let me try to give a hypothetical example if that helps.

Company A wants to monitor their CPU temperature with an Epcos Thermistor (r25 10k ohm, B3974).
Company B wants to monitor the motor temperature with another NTC thermistor (r25 47k ohm, B4050).
A home enthusiast wants to monitor the temperature of their home using an Amertherm thermistor (r25 10k ohm, B3974)

Company B needs to define a resistance to temperature table for their thermistor. Company A and the home enthusiast's thermistor share the same electrical characteristics but use different boards.  They can copy each other's device tree node for the thermistor which is about 90 lines long for defining the table.  If I understand correctly, this is how the UART example you gave works.  Different boards may use the same UART and configure it with different or same clock frequencies and so on.  One line in the DTS configures this for the UART.

What I'm trying to accomplish with the `zephyr_thermistor.dtsi` is make a node defining this 90 line table available to share between boards.  If one of the company's want to swap to another NTC thermistor, they can update one line in their board's dts.

I'm Zephyrizing Linux's implementation of the [NTC Thermistor](https://github.com/torvalds/linux/blob/master/drivers/hwmon/ntc_thermistor.c#L72) but they define the table in C code instead of DTS.   

Does this help clarify?  We can sync up over discord if needed.

mbolivar-nordic replied:
> Does this help clarify and answer your question? We can sync up over discord if needed.

I think we probably better had. We should try to get this into dev-review this week if we can't figure out a solution by EOD wednesday as well, I think.

mbolivar-nordic reviewed (changes requested):
mis-clicked, meant to put -1

pdgendt commented:
I was working on a similar feature based on the linux kernel driver for [`thermal-generic-adc`](https://www.kernel.org/doc/Documentation/devicetree/bindings/thermal/thermal-generic-adc.txt), until I saw this PR. I've created a draft PR of my progress here: #56045 

But it might be too simplistic so I'll close it if this PR gets merged.

I do wonder if the `Epcos b57861s` driver can't be more generic to support multiple types?

cfriedt commented:
> I was working on a similar feature based on the linux kernel driver for [`thermal-generic-adc`](https://www.kernel.org/doc/Documentation/devicetree/bindings/thermal/thermal-generic-adc.txt), until I saw 

Yeah, this looks to be the right way to do it.

In the DT, it you want to be overly specific, you csn have something like

```
compatible = "ellpcos,b57861s0103a039", "zephyr,general-adc-thermistor";
```

In that case, we would intentionally not have an actual binding for `ellpcos,b...` because we would want to to match the generic compatible. 

That level of specificity probably isn't required in all cases though, so the generic compatible (and the instance-specific table) should be enough.

fabiobaltieri commented on dts/bindings/sensor/zephyr_thermistor.dtsi:
```diff
@@ -0,0 +1,98 @@
+		 *   R25 = 10000 ohms
+		 *   B(25/85) = 3974
+		 */
+		sample-datum-0 {
+			temp-c = <(-25)>;
+			resistance = <130370>;
+			sample-index = <0>;
+		};
```
This format seems unnecessarily verbose, how about converting temp-c and resistance to `array`? Then `DT_INST_PROP` would give you an array initializer, just validate that they are the same length with `DT_INST_PROP_LEN` and you'd get away without `num-pairs`, `sample-index` and all this child noise stuff.

asemjonovs committed:
scripts: Create a script to generate an NTC thermistor R/T table

Script to generate an R/T table in device tree format given
R25 and Beta values.

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

mbolivar-nordic commented on dts/bindings/sensor/zephyr,thermistor-rt-table.yaml:
```diff
@@ -0,0 +1,37 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# For more information:
+# https://www.electronics-tutorials.ws/io/thermistors.html
```
This is the sort of thing that belongs better in the `description` since that's in the HTML pages in https://docs.zephyrproject.org/latest/build/dts/api/bindings.html

mbolivar-nordic reviewed (commented):
Looking really good, down to the specifics now! Thanks for the quick update

mbolivar-nordic commented on boards/arm/tdk_robokit1/tdk_robokit1-common.dtsi:
```diff
@@ -6,6 +6,7 @@
 
 #include <zephyr/dt-bindings/adc/adc.h>
 #include "tdk_robokit1-pinctrl.dtsi"
+#include "zephyr_thermistor.dtsi"
```
(Asking out of namespace concern reasons)

asemjonovs commented on scripts/utils/ntc_thermistor_table.py:
```diff
@@ -0,0 +1,96 @@
+"""
+
+import argparse
+import numpy as np
```
Let me see if I can use `import math` so we don't need to rely on `numpy` here

asemjonovs commented on boards/arm/tdk_robokit1/zephyr_thermistor.dtsi:
```diff
@@ -0,0 +1,96 @@
+/ {
+	thermistor_R25_10000_B_3974: thermistor-R25-10000-B-3974 {
+		status = "disabled";
+		compatible = "zephyr,thermistor-rt-table";
```
I wanted to show some association between the temperature and resistance.  Having individual arrays for temperature and another for resistance you lose that association in the dts.  

I tried using a single array (see below) but can't seem to figure out a way to `index+=2` when initializing.  Any suggestions with this method?

```
		/* Format <temp resistance> */
		tr-table = <(-25) 146676>,
		           <(-15) 78875>,
		           <(-5) 44424>,
		           <(5) 26075>,
		           <(15) 15881>,
		           <(25) 10000>,
		           <(35) 6488>,
		           <(45) 4326>,
		           <(55) 2956>,
		           <(65) 2066>,
		           <(75) 1474>,
		           <(85) 1072>,
		           <(95) 793>,
		           <(105) 596>,
		           <(115) 454>,
		           <(125) 351>;
```

Sticking with the child node approach, I can clean-up the redundant `num-pairs` and `sample-index` properties.

cfriedt replied:
Are they both encoded as int32's in dt? If that's case, could just store them as a 1d int32 array and then cast a pointer.

I'll see if I can find the right place to comment on code..

pdgendt replied:
You'll have to create a single array like:
```
		/* Format <temp resistance> */
		tr-table = <(-25) 146676
		           (-15)  78875
		           (-5)   44424
		           (5)    26075
		           (15)   15881
		           (25)   10000
		           (35)   6488
		           (45)   4326
		           (55)   2956
		           (65)   2066
		           (75)   1474
		           (85)   1072
		           (95)   793
		           (105)  596
		           (115)  454
		           (125)  351>;
```

asemjonovs replied:
I updated `struct ntc_compensation` to be both 32 bit values.  I then cast the uint32_t array to a `struct ntc_compensation *`  this lets me keep the association in the device tree.  I attempted to create a union of both pointer types to avoid casting but compiler kept giving `error: initializer element is not constant`.

cfriedt commented on include/zephyr/drivers/zephyr_thermistor.h:
```diff
@@ -0,0 +1,73 @@
+
+#include <zephyr/types.h>
+
+struct ntc_compensation {
```
@asemjonovs - can I suggest just making both of these fields 32-bit? That would make initialization from DT macro 
straightforward.
https://docs.zephyrproject.org/3.1.0/build/dts/bindings.html#property-types

cfriedt replied:
Sorry - just noticed that was already suggested.

pdgendt commented on drivers/sensor/zephyr_thermistor/zephyr_thermistor_device.c:
```diff
@@ -0,0 +1,128 @@
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	adc_sequence_init_dt(&cfg->adc_channel, &sequence);
+	res = adc_read(cfg->adc_channel.dev, &sequence);
+	val_mv = data->raw;
+	res = adc_raw_to_millivolts_dt(&cfg->adc_channel, &val_mv);
+	data->sample_val = val_mv;
```
Validate intermediate results here.

pdgendt commented on drivers/sensor/zephyr_thermistor/zephyr_thermistor_device.c:
```diff
@@ -0,0 +1,128 @@
+	.channel_get = zephyr_thermistor_channel_get,
+};
+
+int zephyr_thermistor_reset(const struct device *dev)
```
`zephyr_thermistor_reset` is only used during in `zephyr_thermistor_init`, move the implementation there and remove this function.

pdgendt commented on drivers/sensor/zephyr_thermistor/zephyr_thermistor_device.c:
```diff
@@ -0,0 +1,128 @@
+	return 0;
+}
+
+int zephyr_thermistor_init(const struct device *dev)
```
```suggestion
static int zephyr_thermistor_init(const struct device *dev)
```

pdgendt commented on scripts/utils/ntc_thermistor_table.py:
```diff
@@ -0,0 +1,96 @@
+
+
+def main(
+    r25: float, beta: float, interval: int, temp_init: float, temp_final: float
```
Why isn't the interval a `float`?

asemjonovs replied:
I refactored so `temp_init`, `temp_final`, and `interval` are all `int`.  I didn't see a need to generate a table with that high of a resolution.  By default it will increase by 10 degree Celsius.

pdgendt commented on drivers/sensor/zephyr_thermistor/zephyr_thermistor_device.c:
```diff
@@ -0,0 +1,128 @@
+                                                                                                   \
+	static const struct zephyr_thermistor_config zephyr_thermistor_cfg_##inst = {              \
+		.adc_channel = ADC_DT_SPEC_INST_GET(inst),                                         \
+		.ntc_cfg =                                                                         \
+			{                                                                          \
+				.r25_ohm = DT_INST_PROP(inst, r25_ohm),                            \
+				.pullup_uv = DT_INST_PROP(inst, pullup_uv),                        \
+				.pullup_ohm = DT_INST_PROP(inst, pullup_ohm),                      \
+				.pulldown_ohm = DT_INST_PROP(inst, pulldown_ohm),                  \
+				.connection_type = DT_INST_STRING_TOKEN(inst, connection_type),    \
+				.type = &NTC_TYPE_NAME(DT_INST_PROP(inst, thermistor)),            \
+			},                                                                         \
```
Nit: indentation
```suggestion
		.ntc_cfg = {									\
			.r25_ohm = DT_INST_PROP(inst, r25_ohm),					\
			.pullup_uv = DT_INST_PROP(inst, pullup_uv),				\
			.pullup_ohm = DT_INST_PROP(inst, pullup_ohm),				\
			.pulldown_ohm = DT_INST_PROP(inst, pulldown_ohm),			\
			.connection_type = DT_INST_STRING_TOKEN(inst, connection_type),		\
			.type = &NTC_TYPE_NAME(DT_INST_PROP(inst, thermistor)),			\
		},										\
```

fabiobaltieri replied:
I think you can fix it with this:
```
/* clang-format off */
/* clang-format on */
```
😉

fabiobaltieri commented on include/zephyr/drivers/zephyr_thermistor.h:
```diff
@@ -0,0 +1,73 @@
+
+#include <zephyr/types.h>
+
+struct ntc_compensation {
+	const int32_t temp_c;
+	const uint32_t ohm;
+};
```
hey does this need to live in `include/zephyr/drivers`? It's only used within the driver, if it should not be exposed to the driver users it should probably be declared inside the driver itself (that is `zephyr_thermistor.c`). You may want to add an extra local header files if you need to share some stuff between the various thermistor files but this should only have the structs part of the "public" API.

asemjonovs replied:
Good catch!

mbolivar-nordic commented on tests/drivers/build_all/sensor/adc.dtsi:
```diff
@@ -0,0 +1,55 @@
+		compatible = "zephyr,thermistor";
+		reg = <0x0 0x1000>;
+		io-channels = <&adc0 0>;
+		thermistor = <&thermistor_R25_10000_B_3974>;
```
The table etc all look great now. I think that's much easier to read personally -- curious what others think.

As one last question (I swear), I find the name of this property a bit odd. This node "is a" thermistor, so it's weird that it also "has a" thermistor. Maybe `zephyr,rt-table` would be a better choice? Or is this patterned off of an existing  binding you're maintaining compatibility with?

MaureenHelm replied:
Agreed, nice work.

I had the same thought on the property name.

asemjonovs replied:
Thanks, I appreciate all the help :-)

teburd reviewed (commented):
Cool that this got even more consolidated into a common thermistor sensor, the community that works :+1:

mbolivar-nordic reviewed (dismissed):
For devicetree.

@asemjonovs @MaureenHelm I'm assuming here that the `zephyr,thermistor` compatible is chosen for future extension into other types of thermistor. (Otherwise, `zephyr,ntc-thermistor` might be a better choice.)

asemjonovs commented:
> For devicetree.
> 
> @asemjonovs @MaureenHelm I'm assuming here that the `zephyr,thermistor` compatible is chosen for future extension into other types of thermistor. (Otherwise, `zephyr,ntc-thermistor` might be a better choice.)

You're right, I'll rename the files to reflect but I'll leave the directory as `zephyr_thermistor` for a future PTC thermistor driver

pdgendt commented on drivers/sensor/zephyr_thermistor/zephyr_ntc_thermistor.c:
```diff
@@ -0,0 +1,150 @@
+
+#define DEFINE_THERMISTOR_TYPE(node_id)                                                            \
+	static int node_id##_compare_ohm(const void *key, const void *element);                    \
+	static const uint32_t node_id##_tr_table[] = DT_PROP(node_id, tr_table);                   \
```
Maybe add a check if the table size if even
```c
	BUILD_ASSERT(DT_PROP_LEN(node_id, tr_table) % 2 == 0,					\
		     "Temperature lookup table needs an even size");				\
```

pdgendt commented on drivers/sensor/zephyr_thermistor/zephyr_ntc_thermistor_device.c:
```diff
@@ -0,0 +1,125 @@
+}
+
+#define ZEPHYR_NTC_THERMISTOR_DEV_INIT(inst)                                                       \
+	static struct zephyr_ntc_thermistor_data zephyr_ntc_thermistor_driver_##inst = {};         \
```
Nit
```suggestion
	static struct zephyr_ntc_thermistor_data zephyr_ntc_thermistor_driver_##inst;         \
```

pdgendt reviewed (commented):
Maybe a dumb question, but doesn't the lookup work for Positive Temperature Coefficient (PTC) thermistors just the same?

pdgendt commented:
> I believe so, for PTC I think the table still needs to be organized in decreasing resistance for the algorithm to be the same. To be honest, I went from knowing nothing about NTC thermistors a month ago to being able to write this driver. I'm still learning :-)

It might be worth renaming to avoid confusion, like `zephyr,adc-thermistor` instead of `zephyr,ntc-thermistor`.

But does would affect a lot of parts of this PR, and I'm no expert either.

mbolivar-nordic commented:
> It might be worth renaming to avoid confusion, like zephyr,adc-thermistor instead of zephyr,ntc-thermistor.

IMO having a more specific compatible is better. You can always use `include` in the bindings to share properties as needed.

mbolivar-nordic reviewed (commented):
The file name should be zephyr,ntc-thermistor.yaml to match the compatible

mbolivar-nordic reviewed (dismissed):
for devicetree

teburd reviewed (dismissed):
lgtm

MaureenHelm reviewed (dismissed):
Looks good, but need to address the twister fail:
```
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) samples/sensor/thermometer/sample.sensor.thermometer on beagleconnect_freedom error (Build failure)
```

asemjonovs commented:
> Looks good, but need to address the twister fail:
> 
> ```
> INFO    - The following issues were found (showing the top 10 items):
> INFO    - 1) samples/sensor/thermometer/sample.sensor.thermometer on beagleconnect_freedom error (Build failure)
> ```

@MaureenHelm 

I don't think my changes affect this.  I'm able to reproduce this error on main branch as well.
```
../../../../../../include/zephyr/device.h:83:41: error: '__device_dts_ord_DT_N_S_sensor_switch_S_hdc2010_humidity_41_BUS_ORD' undeclared here (not in a function); did you mean 'DT_N_S_sensor_switch_S_hdc2010_humidity_41_ORD'?
```

MaureenHelm commented:
> > Looks good, but need to address the twister fail:
> > ```
> > INFO    - The following issues were found (showing the top 10 items):
> > INFO    - 1) samples/sensor/thermometer/sample.sensor.thermometer on beagleconnect_freedom error (Build failure)
> > ```
> 
> @MaureenHelm
> 
> I don't think my changes affect this. I'm able to reproduce this error on main branch as well.
> 
> ```
> ../../../../../../include/zephyr/device.h:83:41: error: '__device_dts_ord_DT_N_S_sensor_switch_S_hdc2010_humidity_41_BUS_ORD' undeclared here (not in a function); did you mean 'DT_N_S_sensor_switch_S_hdc2010_humidity_41_ORD'?
> ```

This has been fixed in main. Please rebase
